vm/ubpf_verifier.c

```diff
@@ -0,0 +1,135 @@
+
+#define _GNU_SOURCE
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdbool.h>
+#include <string.h>
+#include <unistd.h>
+#include <inttypes.h>
+#include <sys/mman.h>
+#include <errno.h>
+#include <assert.h>
+#include "ubpf_int.h"
+#include "ebpf.h"
+
+typedef int (*WALKER)(struct ubpf_vm *vm, struct ebpf_inst inst, void *data, int inst_off, char *visited);
+
+enum ubpf_walk_action
+{
+    UBPF_WALK_CONTINUE,
+    UBPF_WALK_STOP,
+    UBPF_WALK_INVALID,
+};
+
+int isjmp(struct ebpf_inst inst)
+{
+    if (((inst.opcode & EBPF_CLS_MASK) == EBPF_CLS_JMP) && (inst.opcode != EBPF_OP_CALL)) {
+        return true;
+    }
+    return false;
+}
+
+int
+ubpf_walk_paths(struct ubpf_vm *vm, WALKER walk_fn, void *data, int inst_off, char *visited)
+{
+    struct ebpf_inst inst = vm->insts[inst_off];
+    int cmd = walk_fn(vm, inst, data, inst_off, visited);
+    visited[inst_off] = 1;
+    if (cmd != UBPF_WALK_CONTINUE)
+        return cmd;
+    if (inst.opcode == EBPF_OP_EXIT) {
+        return UBPF_WALK_CONTINUE;
+    } else if (isjmp(inst)) {
+        int next_pc = inst_off+1+inst.offset;
+        if (next_pc == inst_off) {
+            fprintf(stderr, "Jump to self at offset %d\n", inst_off);
+            return UBPF_WALK_INVALID;
+        } else if ((next_pc < 0) || (next_pc > vm->num_insts-1)) {
+            fprintf(stderr, "Jump out-of-bounds at offset %d to %d\n", inst_off, next_pc);
+            return UBPF_WALK_INVALID;
+        }
+        if (visited[next_pc] == 0) {
+            cmd = ubpf_walk_paths(vm, walk_fn, data, next_pc, visited);
+            if (cmd == UBPF_WALK_STOP || cmd == UBPF_WALK_INVALID)
+                return cmd;
+        }
+    }
+    if (inst_off == vm->num_insts-1) {
+        return UBPF_WALK_CONTINUE;
+    } else {
+        return ubpf_walk_paths(vm, walk_fn, data, inst_off+1, visited);
+    }
+}
+
+int
+ubpf_walk_start(struct ubpf_vm *vm, WALKER walk_fn, void *data)
+{
+    char visited[vm->num_insts];
+    memset((void *)visited, 0, vm->num_insts);
+    return ubpf_walk_paths(vm, walk_fn, data, 0, visited);
+}
+
+int
+_walker_no_dead_insts(struct ubpf_vm *vm, struct ebpf_inst inst, void *data, int inst_off, char *visited)
+{
+    return UBPF_WALK_CONTINUE;
+}
+
+int
+ubpf_verify_no_dead_insts(struct ubpf_vm *vm)
+{
+    char visited[vm->num_insts];
+    memset((void *)visited, 0, vm->num_insts);
+    int ret = ubpf_walk_paths(vm, _walker_no_dead_insts, NULL, 0, visited);
+    if (ret)
+        return ret;
+    int any_dead = 0;
+    for (int i = 0; i < vm->num_insts; i++) {
+        if (visited[i] == 0) {
+            any_dead = 1;
+            fprintf(stderr, "Dead instruction at offset %d\n", i);
+        }
+    }
+    return any_dead;
+}
+
+int
+_walker_no_loops(struct ubpf_vm *vm, struct ebpf_inst inst, void *data, int inst_off, char *visited)
+{
+    if (isjmp(inst) && (inst_off+1+inst.offset < inst_off) && visited[inst_off+1+inst.offset]) {
+        fprintf(stderr, "Loop detected at offset %d\n", inst_off);
```

Comment:
Good catch! To fix this, I suspect that it would probably be best to construct a CFG so that we can reason about blocks which branch to a predecessor block, right? So then, to determine if we have a loop, we check if we can get back to our current block by following each edge exiting our block. Does this sound reasonable, and if so, should I start constructing a CFG?

Reply:
Maybe it would be beneficial to describe the algorithm here before diving into its implementation? What did you have in mind?

Last time I had to implement an algorithm to find loops ([for Oko](https://github.com/Orange-OpenSource/oko/blob/599e0a2706988486498d51a97a7e673ad2bc8fb7/lib/bpf/ubpf_vm.c#L1442), Apache2 license), I only needed to annotate the vertices and edges IIRC. It seems [DPDK](https://github.com/DPDK/dpdk/blob/1f1619994570210105ea9fa674562c11ce1921b3/lib/librte_bpf/bpf_validate.c#L1961) has a different approach which we may also use (BSD license AFAICT).

Reply:
I was thinking that for each block, if you walk along each path leaving the block, for any path you take, you'll either end up at a program terminator (`exit`), or back at the same block. But how to do that efficiently (instead of doing a full walk for every block), we should probably copy from somewhere else.

Since it seems like you have a full copy of ubpf in oko, how would you feel about me copying the relevant loop verifier bits from oko? It would be great if you didn't need to maintain two copies of ubpf then :smile: